File: app/src/main/java/com/hippo/ehviewer/dao/DownloadInfo.java
package com.hippo.ehviewer.dao;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;

import android.os.Parcel;
import com.hippo.ehviewer.client.data.GalleryInfo;
import com.google.gson.JsonObject;
import java.util.ArrayList;

/**
 * Entity mapped to table "DOWNLOADS".
 * Primary key is GID (inherited from GalleryInfo).
 */
@Entity(tableName = "DOWNLOADS", primaryKeys = {"GID"})
public class DownloadInfo extends GalleryInfo {

	@ColumnInfo(name = "STATE") public int state;
	@ColumnInfo(name = "LEGACY") public int legacy;
	@ColumnInfo(name = "TIME") public long time;
	@androidx.annotation.Nullable @ColumnInfo(name = "LABEL") public String label;
	@androidx.annotation.Nullable @ColumnInfo(name = "ARCHIVE_URI") public String archiveUri;

	public static final Creator<DownloadInfo> CREATOR = new Creator<DownloadInfo>() {
		@Override
		public DownloadInfo createFromParcel(Parcel source) {
			return new DownloadInfo(source);
		}

		@Override
		public DownloadInfo[] newArray(int size) {
			return new DownloadInfo[size];
		}
	};
	public static final int STATE_INVALID = -1;
	public static final int STATE_NONE = 0;
	public static final int STATE_WAIT = 1;
	public static final int STATE_DOWNLOAD = 2;
	public static final int STATE_FINISH = 3;
	public static final int STATE_FAILED = 4;
	public static final int STATE_UPDATE = 5;
	public static final int GOTO_NEW = 6;
	@Ignore public long speed;
	@Ignore public long remaining;
	@Ignore public int finished;
	@Ignore public int downloaded;
	@Ignore public int total;
	@Ignore public long fileSize = -1;


	public DownloadInfo() {
	}

	@Ignore
	public DownloadInfo(long gid) {
		this.gid = gid;
	}

	@Ignore
	public DownloadInfo(long gid, String token, String title, String titleJpn, String thumb, int category,
			String posted, String uploader, float rating, String simpleLanguage, int state, int legacy, long time,
			String label, String archiveUri) {
		this.gid = gid;
		this.token = token;
		this.title = title;
		this.titleJpn = titleJpn;
		this.thumb = thumb;
		this.category = category;
		this.posted = posted;
		this.uploader = uploader;
		this.rating = rating;
		this.simpleLanguage = simpleLanguage;
		this.state = state;
		this.legacy = legacy;
		this.time = time;
		this.label = label;
		this.archiveUri = archiveUri;
	}

	public long getGid() { return gid; }
	public void setGid(long gid) { this.gid = gid; }
	public String getToken() { return token; }
	public void setToken(String token) { this.token = token; }
	public String getTitle() { return title; }
	public void setTitle(String title) { this.title = title; }
	public String getTitleJpn() { return titleJpn; }
	public void setTitleJpn(String titleJpn) { this.titleJpn = titleJpn; }
	public String getThumb() { return thumb; }
	public void setThumb(String thumb) { this.thumb = thumb; }
	public int getCategory() { return category; }
	public void setCategory(int category) { this.category = category; }
	public String getPosted() { return posted; }
	public void setPosted(String posted) { this.posted = posted; }
	public String getUploader() { return uploader; }
	public void setUploader(String uploader) { this.uploader = uploader; }
	public float getRating() { return rating; }
	public void setRating(float rating) { this.rating = rating; }
	public String getSimpleLanguage() { return simpleLanguage; }
	public void setSimpleLanguage(String simpleLanguage) { this.simpleLanguage = simpleLanguage; }
	public int getState() { return state; }
	public void setState(int state) { this.state = state; }
	public int getLegacy() { return legacy; }
	public void setLegacy(int legacy) { this.legacy = legacy; }
	public long getTime() { return time; }
	public void setTime(long time) { this.time = time; }
	public String getLabel() { return label; }
	public void setLabel(String label) { this.label = label; }
	public String getArchiveUri() { return archiveUri; }
	public void setArchiveUri(String archiveUri) { this.archiveUri = archiveUri; }

	@Override
	public int describeContents() { return 0; }

	@Override
	public void writeToParcel(Parcel dest, int flags) {
		super.writeToParcel(dest, flags);
		dest.writeInt(this.state);
		dest.writeInt(this.legacy);
		dest.writeLong(this.time);
		dest.writeString(this.label);
		dest.writeString(this.archiveUri);
	}

	@Ignore
	protected DownloadInfo(Parcel in) {
		super(in);
		this.state = in.readInt();
		this.legacy = in.readInt();
		this.time = in.readLong();
		this.label = in.readString();
		this.archiveUri = in.readString();
	}

	@Ignore
	public DownloadInfo(GalleryInfo galleryInfo) {
		this.gid = galleryInfo.gid;
		this.token = galleryInfo.token;
		this.title = galleryInfo.title;
		this.titleJpn = galleryInfo.titleJpn;
		this.thumb = galleryInfo.thumb;
		this.category = galleryInfo.category;
		this.posted = galleryInfo.posted;
		this.uploader = galleryInfo.uploader;
		this.rating = galleryInfo.rating;
		this.simpleTags = galleryInfo.simpleTags;
		this.simpleLanguage = galleryInfo.simpleLanguage;
		this.serverProfileId = galleryInfo.serverProfileId;
	}

	public void updateInfo(GalleryInfo galleryInfo) {
		this.token = galleryInfo.token;
		this.title = galleryInfo.title;
		this.titleJpn = galleryInfo.titleJpn;
		this.thumb = galleryInfo.thumb;
		this.category = galleryInfo.category;
		this.posted = galleryInfo.posted;
		this.uploader = galleryInfo.uploader;
		this.rating = galleryInfo.rating;
		this.simpleTags = galleryInfo.simpleTags;
		this.simpleLanguage = galleryInfo.simpleLanguage;
	}

	public JsonObject toJson() {
		JsonObject jsonObject = super.toJson();
		jsonObject.addProperty("finished", finished);
		jsonObject.addProperty("legacy", legacy);
		jsonObject.addProperty("label", label);
		jsonObject.addProperty("downloaded", downloaded);
		jsonObject.addProperty("remaining", remaining);
		jsonObject.addProperty("speed", speed);
		jsonObject.addProperty("state", state);
		jsonObject.addProperty("time", time);
		jsonObject.addProperty("total", total);
		jsonObject.addProperty("archiveUri", archiveUri);
		return jsonObject;
	}

	public static DownloadInfo downloadInfoFromJson(JsonObject object) throws ClassCastException {
		DownloadInfo downloadInfo = (DownloadInfo) GalleryInfo.galleryInfoFromJson(object);
		downloadInfo.finished = object.has("finished") ? object.get("finished").getAsInt() : 0;
		downloadInfo.legacy = object.has("legacy") ? object.get("legacy").getAsInt() : 0;
		downloadInfo.label = object.has("label") ? object.get("label").getAsString() : null;
		downloadInfo.downloaded = object.has("downloaded") ? object.get("downloaded").getAsInt() : 0;
		downloadInfo.remaining = object.has("remaining") ? object.get("remaining").getAsLong() : 0;
		downloadInfo.speed = object.has("speed") ? object.get("speed").getAsLong() : 0;
		downloadInfo.state = object.has("state") ? object.get("state").getAsInt() : 0;
		downloadInfo.time = object.has("time") ? object.get("time").getAsLong() : 0;
		downloadInfo.total = object.has("total") ? object.get("total").getAsInt() : 0;
		downloadInfo.archiveUri = object.has("archiveUri") ? object.get("archiveUri").getAsString() : null;
		return downloadInfo;
	}

}

File: app/src/main/java/com/hippo/ehviewer/dao/DownloadInfo.kt
/*
 * Copyright 2018 Hippo Seven
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hippo.ehviewer.dao

import android.os.Parcel
import android.os.Parcelable
import com.hippo.ehviewer.download.DownloadState
import com.lanraragi.reader.domain.Archive

/**
 * In-memory view over a `download` row from `ARCHIVE_LOCAL_STATE`.
 *
 * Post-L1-4 this class is no longer a Room `@Entity` — the persisted
 * state lives on [ArchiveLocalState] (one row per arcid, with a
 * download subsystem when DOWNLOAD_STATE is non-null). UI / Adapter /
 * sync code keeps reading and writing `DownloadInfo` instances; the
 * repository layer translates between this view and the unified row.
 *
 * Mutable fields (state, legacy, label, archiveUri, time, etc.) are
 * preserved because the W36-era listeners and adapters mutate this
 * object in place. Demoting to an immutable data class would force a
 * larger UI refactor; that work belongs to a follow-up.
 *
 * Parcelable is kept because [DownloadInfo] still flows through Intent
 * extras in a few places (DownloadService, GalleryActivity).
 */
class DownloadInfo() : Parcelable {

    // ── Display fields (from Archive payload) ──

    @JvmField
    var arcid: String = ""

    @JvmField
    var title: String? = null

    @JvmField
    var thumb: String? = null

    @JvmField
    var rating: Float = 0f

    @JvmField
    var simpleLanguage: String? = null

    @JvmField
    var serverProfileId: Long = 0

    // ── Download subsystem fields ──

    @JvmField
    var state: DownloadState = DownloadState.NONE

    @JvmField
    var legacy: Int = 0

    @JvmField
    var time: Long = 0

    @JvmField
    var label: String? = null

    @JvmField
    var archiveUri: String? = null

    /**
     * SAF tree URI of the download root in effect when this archive was
     * downloaded. Mirrored from `ARCHIVE_LOCAL_STATE.DOWNLOAD_ROOT_URI`.
     * NULL means "use the current `DownloadSettings.getDownloadLocation()`"
     * — applies to legacy rows before the v25→v26 boot backfill runs,
     * and to future rows where the user cleared the download location
     * entirely. Read paths (gallery viewer, resume worker, delete) prefer
     * this value over the current setting so old downloads stay
     * reachable after the user changes the setting.
     */
    @JvmField
    var downloadRootUri: String? = null

    // ── Transient, non-persisted helpers ──

    /**
     * Display tags; populated by the repository layer from the
     * decoded Archive's flat tag list.
     */
    @JvmField
    var simpleTags: Array<String>? = null

    /**
     * Search-side tag list, read by
     * [com.hippo.ehviewer.sync.DownloadListInfosExecutor.matchTag].
     */
    @JvmField
    var tgList: ArrayList<String>? = null

    /**
     * Cached size of the on-disk download directory; filled lazily by
     * [com.hippo.ehviewer.sync.DownloadListInfosExecutor] when sorting
     * by size. Not progress data — separate from
     * `DownloadProgressTracker` (see ADR-001).
     */
    @JvmField
    var fileSize: Long = -1

    // ── Parcelable ──

    private constructor(`in`: Parcel) : this() {
        arcid = `in`.readString() ?: ""
        title = `in`.readString()
        thumb = `in`.readString()
        rating = `in`.readFloat()
        simpleLanguage = `in`.readString()
        serverProfileId = `in`.readLong()
        state = DownloadState.fromCode(`in`.readInt())
        legacy = `in`.readInt()
        time = `in`.readLong()
        label = `in`.readString()
        archiveUri = `in`.readString()
        downloadRootUri = `in`.readString()
        simpleTags = `in`.createStringArray()
        @Suppress("UNCHECKED_CAST")
        tgList = `in`.readArrayList(String::class.java.classLoader) as? ArrayList<String>
        fileSize = `in`.readLong()
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(arcid)
        dest.writeString(title)
        dest.writeString(thumb)
        dest.writeFloat(rating)
        dest.writeString(simpleLanguage)
        dest.writeLong(serverProfileId)
        dest.writeInt(state.code)
        dest.writeInt(legacy)
        dest.writeLong(time)
        dest.writeString(label)
        dest.writeString(archiveUri)
        dest.writeString(downloadRootUri)
        dest.writeStringArray(simpleTags)
        dest.writeList(tgList)
        dest.writeLong(fileSize)
    }

    /**
     * Refresh display fields from a re-fetched [Archive]. Called from
     * GalleryDetailScene after detail load to keep the cached download
     * row in sync with the latest server state. Note that this only
     * mutates the in-memory view; persistence happens via
     * [com.hippo.ehviewer.dao.DownloadDbRepository.putDownloadInfo].
     *
     * simpleLanguage is intentionally not refreshed because LRR never
     * populates it; the column has no producer in the post-L1 schema.
     */
    fun updateInfo(archive: Archive) {
        arcid = archive.arcid
        title = archive.title
        thumb = archive.thumbnailUrl
        rating = archive.rating
        // simpleTags is recomputed by the adapter from the live Archive
        // on the next bind, so no copy is needed here.
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<DownloadInfo> = object : Parcelable.Creator<DownloadInfo> {
            override fun createFromParcel(source: Parcel): DownloadInfo = DownloadInfo(source)
            override fun newArray(size: Int): Array<DownloadInfo?> = arrayOfNulls(size)
        }
    }
}
